pkg/master/thirdparty_controller_test.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package master
+
+import (
+	"testing"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/expapi"
+	"k8s.io/kubernetes/pkg/registry/thirdpartyresourcedata"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+type FakeAPIInterface struct {
+	removed   []string
+	installed []*expapi.ThirdPartyResource
+	apis      []string
+	t         *testing.T
+}
+
+func (f *FakeAPIInterface) RemoveThirdPartyResource(path string) error {
+	f.removed = append(f.removed, path)
+	return nil
+}
+
+func (f *FakeAPIInterface) InstallThirdPartyResource(rsrc *expapi.ThirdPartyResource) error {
+	f.installed = append(f.installed, rsrc)
+	_, group, _ := thirdpartyresourcedata.ExtractApiGroupAndKind(rsrc)
+	f.apis = append(f.apis, makeThirdPartyPath(group))
+	return nil
+}
+
+func (f *FakeAPIInterface) HasThirdPartyResource(rsrc *expapi.ThirdPartyResource) (bool, error) {
+	if f.apis == nil {
+		return false, nil
+	}
+	_, group, _ := thirdpartyresourcedata.ExtractApiGroupAndKind(rsrc)
+	path := makeThirdPartyPath(group)
+	for _, api := range f.apis {
+		if api == path {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (f *FakeAPIInterface) ListThirdPartyResources() []string {
+	return f.apis
+}
+
+func TestSyncAPIs(t *testing.T) {
+	tests := []struct {
+		list              *expapi.ThirdPartyResourceList
+		apis              []string
+		expectedInstalled []string
+		expectedRemoved   []string
+		name              string
+	}{
+		{
+			list: &expapi.ThirdPartyResourceList{
+				Items: []expapi.ThirdPartyResource{
+					{
+						ObjectMeta: api.ObjectMeta{
+							Name: "foo.example.com",
+						},
+					},
+				},
+			},
+			expectedInstalled: []string{"foo.example.com"},
+			name:              "simple add",
+		},
+		{
+			list: &expapi.ThirdPartyResourceList{
+				Items: []expapi.ThirdPartyResource{
+					{
+						ObjectMeta: api.ObjectMeta{
+							Name: "foo.example.com",
+						},
+					},
+				},
+			},
+			apis: []string{
+				"/thirdparty/example.com",
+				"/thirdparty/example.com/v1",
```

Comment:
No, we add two webservices for every API in the master, this is the expected thing for the master to return.